apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/Tasks/TaskAwaitable.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.PubSub.V1.Tasks
+{
+    internal struct TaskAwaitable
+    {
+        private class Awaiter : ITaskAwaiter
+        {
+            public Awaiter(Task task, TaskScheduler taskScheduler)
+            {
+                _task = task;
+                _taskScheduler = taskScheduler;
+            }
+
+            private Task _task;
+            private TaskScheduler _taskScheduler;
+
+            public void OnCompleted(Action continuation) => _task.ContinueWith(_ => continuation(), _taskScheduler);
+            public void UnsafeOnCompleted(Action continuation) => _task.ContinueWith(_ => continuation(), _taskScheduler);
```

Comment:
This specific nested class is only used for testing, so is almost certainly moveable to a test[ing] package.

Unsafe/non-unsafe is about propagating ExecutionContext (unsafe doesn't need to).
As I understand it, because this class implements ICriticalNotifyCompletion (Via ITaskAwaiter) the OnCompleted() method will never be called. But I could be wrong.
If this is moved out of prod, then only the forwarding awaiter needs to remain in prod, which defines no new behaviour.

I'm fairly strongly of the opinion that the testing that this enables is worth this code. It adds ~zero complexity to the actual code.